clients/drcachesim/tools/opcode_mix.cpp

```diff
@@ -222,72 +213,27 @@ opcode_mix_t::parallel_shard_memref(void *shard_data, const memref_t &memref)
-        decode_pc = const_cast<app_pc>(memref.instr.encoding);
-        if (memref.instr.encoding_is_new) {
-            // The code may have changed: invalidate the cache.
-            shard->worker->opcode_data_cache.erase(trace_pc);
-        }
-    } else {
-        // Legacy trace support where we need the binaries.
-        if (!module_mapper_) {
-            shard->error =
-                "Module file path is missing and trace has no embedded encodings";
-            return false;
-        }
-        if (trace_pc >= shard->last_trace_module_start &&
-            static_cast<size_t>(trace_pc - shard->last_trace_module_start) <
-                shard->last_trace_module_size) {
-            decode_pc = shard->last_mapped_module_start +
-                (trace_pc - shard->last_trace_module_start);
-        } else {
-            std::lock_guard<std::mutex> guard(mapper_mutex_);
-            decode_pc = module_mapper_->find_mapped_trace_bounds(
-                trace_pc, &shard->last_mapped_module_start,
-                &shard->last_trace_module_size);
-            if (!module_mapper_->get_last_error().empty()) {
-                shard->last_trace_module_start = nullptr;
-                shard->last_trace_module_size = 0;
-                shard->error = "Failed to find mapped address for " +
-                    to_hex_string(memref.instr.addr) + ": " +
-                    module_mapper_->get_last_error();
-                return false;
-            }
-            shard->last_trace_module_start =
-                trace_pc - (decode_pc - shard->last_mapped_module_start);
-        }
-    }
-    int opcode;
-    uint category;
-    auto cached_opcode_category = shard->worker->opcode_data_cache.find(trace_pc);
-    if (cached_opcode_category != shard->worker->opcode_data_cache.end()) {
-        opcode = cached_opcode_category->second.opcode;
-        category = cached_opcode_category->second.category;
-    } else {
-        instr_t instr;
-        instr_init(dcontext_.dcontext, &instr);
-        app_pc next_pc =
-            decode_from_copy(dcontext_.dcontext, decode_pc, trace_pc, &instr);
-        if (next_pc == NULL || !instr_valid(&instr)) {
-            instr_free(dcontext_.dcontext, &instr);
-            shard->error =
-                "Failed to decode instruction " + to_hex_string(memref.instr.addr);
-            return false;
-        }
-        opcode = instr_get_opcode(&instr);
-        category = instr_get_category(&instr);
-        shard->worker->opcode_data_cache[trace_pc] = opcode_data_t(opcode, category);
-        instr_free(dcontext_.dcontext, &instr);
+    shard->error = shard->decode_cache->add_decode_info(memref.instr);
+    if (shard->error != "") {
+        return false;
     }
-    ++shard->opcode_counts[opcode];
-    ++shard->category_counts[category];
+    opcode_data_t *opcode_data =
+        shard->decode_cache->get_decode_info(reinterpret_cast<app_pc>(memref.instr.addr));
```

Comment:
But now we have to do two lookups: one for the add and one here.  Better to do just one lookup for the common case of the instr already being there.

Reply:
Done.

In a previous version I made add_decode_info also return the decode info object, but that seemed to just make some things more complex so I removed it. Just doing a get_decode_info before the add_decode_info in the user is good though.

Reply:
Actually, we need to add_decode_info on each instr, in case encoding_is_new and the cached info needs to be replaced.

If we really want to optimize the two lookups, we can have add_decode_info also return the cached info using a reference arg. Like I noted above, when I tried that there were some complexities.
- if decoding fails, currently add_decode_info returns an error, and subsequent get_decode_info returns the default-constructed decode_info (the one that returns is_valid()=false). If we make add_decode_info return the cached info, in the above case, it will return an error string and also return the cached info, which may be confusing for a user.
- if the output reference arg is a non-ptr object, we'll be creating a copy of the cached info in every add_decode_info, which is also overhead. If the output reference arg is a ptr object, it will never be nullptr, but then a user might wonder why it is a ptr.

So, it seemed cleaner initially to me to just have the two separate APIs. unordered_map lookup is constant time (though I don't have benchmarks) so maybe it's fine to keep that, for a simpler API?

Reply:
Actually, I guess the contract isn't any more confusing than what get_decode_info already does. I'll make the change to have add_decode_info return the cached_decode_info.

Reply:
Done. Also optimized away other redundant lookups during add_decode_info.